src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -192,6 +196,33 @@ partitions_request_iterator validate_partitions(
     return valid_partitions_end;
 }
 
+ss::future<std::optional<cluster::topic_metadata>>
+get_topic_metadata_with_retry(
+  const cluster::metadata_cache& md_cache, model::topic topic) {
+    auto retries
+      = config::shard_local_cfg().metadata_dissemination_retries.value();
+    auto delay_ms
+      = config::shard_local_cfg().metadata_dissemination_retry_delay_ms.value();
+    std::optional<cluster::topic_metadata> tp_metadata;
+    model::topic_namespace tp_ns{model::kafka_namespace, topic};
+
+    while (!tp_metadata.has_value() && 0 < retries--) {
+        vlog(
+          klog.debug,
+          "Fetching topic metadata: ns {}, retries {}",
+          tp_ns,
+          retries);
+        tp_metadata = md_cache.get_topic_metadata(tp_ns);
+        // Only sleep if the metadata fetch failed to avoid any impact to
+        // latency
+        if (!tp_metadata.has_value()) {
```

Comment:
I guess theres no way around this, but in the event the client passes in a topic that doesn't exist they will have to wait the max interval to receive a response for every topic in the request.

I still feel somewhat weird about this, since no other request handler in the codebase does this, and yet they all perform the same type of validation.

Reply:
> I still feel somewhat weird about this, since no other request handler in the codebase does this, and yet they all perform the same type of validation.

What do you mean by "same type of validation?" No other Kafka handler/request in our codebase has the ability to change a topic's replication factor. So, IMO, this check and retry loop are unique to AlterPartitionsReassignments.
Also, let's not forget that the reason we need `topic_metadata` is to stop users from upreplicating/dereplicating a topic. So, once we resolve that issue then this check and retry loop can go away.

Reply:
"same type of validation" meaning, checking for things like the number of partitions in a topic. An example of that can be seen [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/handlers/create_partitions.cc#L196) in the create_partitions request handler.

If this is indeed an issue with questioning the validity of topic_metadata, then yes every request would have to poll on `tp_metadata.has_value()` after accessing it.

Reply:
@graphcareful 

> "same type of validation" meaning, checking for things like the number of partitions in a topic, or anything else for that matter that uses the result from `get_topic_metadata`. An example of a request that looks for the number of partitions for validation purposes can be seen [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/handlers/create_partitions.cc#L196) in the create_partitions request handler.

I don't know if that's the best example. In that handler we are assessing the partition count from a `topic_partition`. I was told that the replication factor field from `topic_partition` is deprecated so I wonder if partition count is also deprecated https://github.com/redpanda-data/redpanda/pull/7685#discussion_r1081040543

> If this is indeed an issue with questioning the validity of topic_metadata, then yes every request would have to poll on `tp_metadata.has_value()` after accessing it.

Fair point. Let's move the discussion to slack for now and I'll see about setting up a quick sync with the replication team so we can resolve this Monday.

A possible alternative is instead return a retriable error in AlterPartitions and have the DT test retry the request. The only problem here is that the Kafka errors that semantically make sense for a check on replication factor are **not** retriable errors (e.g., invalid_replication_factor)

Reply:
> One of the validators in the AlterPartitionReassignments API relies on `topic_metadata` to make sure the incoming requests does not change the topic replication factor. However, `topic_metadata` could be undefined due to eventual consistency in the `topic_table`. This PR adds a retry loop to `topic_metadata` fetch as well as more logging.

From sync via video chat, we decided that the issue and solution is instead:

One of the validators in the AlterPartitionReassignments API relies on `topic_metadata` to make sure the incoming requests does not change the topic replication factor. However, `topic_metadata` could be stale because the broker does not yet have updated metadata but it will soon. The solution is to return a retriable error and add retry loops to tests.